src/bundles/peer-locations.js

```diff
@@ -300,44 +138,74 @@ class PeerLocationResolver {
       ...opts.cache
     })
 
+    this.failedAddrs = []
+
+    this.queue = queue({
+      concurrency: opts.concurrency,
+      autostart: true
+    })
+
     this.geoipLookupPromises = {}
   }
 
-  async lookupWithCaches (peerId, addr, getIpfs) {
-    const ipv4Tuple = Multiaddr(addr).stringTuples().find(isNonHomeIPv4)
-    if (!ipv4Tuple) {
-      throw new Error(`Unable to resolve location for non-IPv4 address ${addr}`)
-    }
+  async findLocations (peers, getIpfs) {
+    const res = {}
 
-    var ipv4Addr = ipv4Tuple[1]
+    for (const p of peers) {
+      const peerId = p.peer.toB58String()
+      const addr = p.addr.toString()
 
-    // maybe we have it cached by ipv4 address already, check that.
-    let location = await this.geoipCache.get(ipv4Addr)
-    if (location) {
-      return location
-    }
+      const ipv4Tuple = Multiaddr(addr).stringTuples().find(isNonHomeIPv4)
+      if (!ipv4Tuple) {
+        continue
+      }
 
-    // no ip address cached. are we looking it up already?
-    var locPromise = this.geoipLookupPromises[ipv4Addr]
+      const ipv4Addr = ipv4Tuple[1]
+      if (this.failedAddrs.includes(ipv4Addr)) {
+        continue
+      }
 
-    if (!locPromise) {
-      locPromise = new Promise((resolve, reject) => {
-        const ipfs = getIpfs()
-        geoip.lookup(ipfs, ipv4Addr, (err, data) => {
-          // save it in the caches, and remove the promise.
-          this.geoipCache.set(ipv4Addr, data)
-          delete this.geoipLookupPromises[ipv4Addr]
+      // maybe we have it cached by ipv4 address already, check that.
+      const location = await this.geoipCache.get(ipv4Addr)
+      if (location) {
+        res[peerId] = location
+        continue
+      }
+
+      // no ip address cached. are we looking it up already?
+      var locPromise = this.geoipLookupPromises[ipv4Addr]
+      if (locPromise) {
+        continue
+      }
+
+      this.queue.push(async () => {
+        return new Promise((resolve, reject) => {
+          const ipfs = getIpfs()
+
+          geoip.lookup(ipfs, ipv4Addr, (err, data) => {
+            delete this.geoipLookupPromises[ipv4Addr]
+
+            if (err) {
+              // mark this one as failed so we don't retry again
+              this.failedAddrs.push(ipv4Addr)
 
-          if (err) return reject(err)
-          resolve(data)
+              // if there's 500 or more failed addresses, remove the
+              // oldest 100 failures.
+              if (this.failedAddrs.length >= 500) {
+                this.failedAddrs.splice(0, 100)
+              }
```

Comment:
Small nit: drop-ins like this, even with good comments like this one,  make code harder to reason about :P
Let's remove this custom code by using [hashlru](https://www.npmjs.com/package/hashlru) for `failedAddrs`.
It comes with known characteristics, removes cognitive overhead (new contributor sees LRU and knows what is happening), and we already use it in js-ipfs.